lib/discord/client.py

```diff
@@ -0,0 +1,127 @@
+from lib.discord.op.command.abst_command_base import AbstCommandBase
+from lib.discord.op.command.command_register import CommandRegister
+from lib.discord.op.event import approve_event
+from lib.discord.op.event.reaction_event import ReactionEvent
+from lib.logging.logger import log
+from lib.settings.setting import Setting
+
+
+class MainClient(discord.Client):
+    """
+    DiscordのBot。
+    """
+
+    def __init__(self, setting: Setting, vote_record: TweetsVoteRecord):
+        """
+        指定した設定でクライアントを初期化する。
+        :param setting: Botに使用する設定。
+        :param vote_record: ツイートの投票を記録するレコード。
+        """
+        super(MainClient, self).__init__()
+        self.reaction_event_handler: ReactionEvent = ReactionEvent(setting, vote_record)
+        self.command_register: CommandRegister = CommandRegister(setting)
+        self.setting: Setting = setting
+        self.vote_record: TweetsVoteRecord = vote_record
+        self.activity_channel: discord.TextChannel = None
+
+    def launch(self):
+        """
+        Botを起動する。
+        Botが何らかの理由で終了するまで処理は停止する。
+        """
+        log("client-login", "ログイン処理を開始します。")
+        self.run(self.setting.token)
+
+    def add_command(self, command: Type[AbstCommandBase]):
+        """
+        実行対象のコマンドを登録する。
+        :param command: 実行対象のコマンドのType。
+        """
+        self.command_register.add_command(command)
+
+    async def on_ready(self):
+        """
+        Botが起動した際に呼ばれる。
+        必要なチャンネルが存在するかを確認し、コマンドの初期化を行う。
+        """
+        log("client-login", "ログインに成功しました。適切な設定が行われているか確認しています。")
+        self.activity_channel = self.get_channel(self.setting.activity_channel_id)
+
+        if self.activity_channel is None:
+            log("client-login", "アクティビティチャンネルが見つかりません。設定に異常があります。")
+            raise RuntimeError("Activity channel is not found! Check your \"activity_channel_id\" value.")
+
+        log("client-login", "設定に問題はありませんでした。コマンドのインスタンスを生成します…")
+        self.command_register.initialize_commands(self.activity_channel.guild, self.vote_record)
+
+        log("client-login", "問題は発生しませんでした。起動メッセージを送信します…")
+        await self.activity_channel.send("***†Delitter Ready†***")
+
+    async def on_message(self, message: discord.Message):
+        """
+        メッセージを受信したときに発火されるイベント。
+        処理対象のメッセージであるかを確認し、適切なコマンドを実行する。
+        :param message: 受信したメッセージの情報。
+        """
+
+        # 処理対象のメッセージかを確認する
+        if message.author.bot or message.channel.id != self.setting.activity_channel_id:
+            log("client-msg", "メッセージはBotからのものか、Activity Channelではないところで発言されたものでした。無視します！")
+            return
+
+        if not message.content.startswith(self.setting.prefix):
+            log("client-msg", "メッセージは処理対象でしたが、Prefix「{}」がありませんでした。無視します！".format(self.setting.prefix))
+            return
+
+        log("client-msg", "処理対象のメッセージを受信しました:\n{}".format(message.content))
+
+        # コマンドをパースする
+        await self.command_register.execute_command(message)
+
```

Comment:
try-except and error notify(like send error to user) is needed because users love breaking bot.

Reply:
265f485